Update budget recommendation API to support multiple countries

ianlin opened the pull request:
### Changes proposed in this Pull Request:

Part of #1248. This PR updates the API `ads/campaigns/budget-recommendation/:country-code` to `ads/campaigns/budget-recommendation?country_codes=` in order to support multiple countries' budget recommendation.

### Screenshots

<img width="820" alt="Screenshot 2022-02-18 at 17 52 53" src="https://user-images.githubusercontent.com/914406/154659436-2b1915c6-c4fc-4b0c-b82f-c9849070bb87.png">

### Detailed test instructions:

Since the frontend hasn't integrated with this new API to show the budget recommendation for multiple countries, one can test this PR through any HTTP clients like Postman.

#### Request

```
GET https://testdomain/wp-json/wc/gla/ads/campaigns/budget-recommendation?country_codes=TW,GB,US
```

#### Response

```json
{
    "currency": "EUR",
    "recommendations": [
        {
            "country": "US",
            "daily_budget_low": 10,
            "daily_budget_high": 25
        },
        {
            "country": "GB",
            "daily_budget_low": 5,
            "daily_budget_high": 20
        },
        {
            "country": "IE",
            "daily_budget_low": 5,
            "daily_budget_high": 15
        }
    ]
}
```

### Additional details

The type of the budget in `daily_budget_low` and `daily_budget_high` was of `string`, I changed the type to `number` in order to align with the type of the response data `amount` when calling `ads/campaigns` API.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Update - Budget recommendation for multiple countries

ianlin committed:
Update API ads/campaigns/budget-recommendation to support multi-country

ianlin committed:
Add unit tests for budget recommendation controller

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -113,24 +142,24 @@ protected function get_schema_properties(): array {
 				'sanitize_callback' => $this->get_country_code_sanitize_callback(),
 				'validate_callback' => $this->get_country_code_validate_callback(),
 				'required'          => true,
 			],
```
I think we should try and use a valid type for this, since it's an array of strings you would define that as shown here: https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#arrays

ianlin replied:
Thanks for the info. Updated in eb0425b to use a new schema for query parameters.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
-			$country        = strtoupper( $request->get_param( 'country_code' ) );
-			$currency       = $this->middleware->get_ads_currency();
-			$recommendation = $this
+			$country_codes = $request->get_query_params()['country_codes'] ?? '';
```
If we are getting one specific parameter I think it's better to use `get_param( 'country_codes' )`. We have already registered this as a GET route so it will always have to be a query parameter.

WordPress has built in functionality for validating the Rest API parameters, so we don't need to default to empty here. Although that's not happening in this controller, because when we register the route we don't provide the `args`, you can see an example of how that's done [here](https://github.com/woocommerce/google-listings-and-ads/blob/1.11.1/src/API/Site/Controllers/Ads/CampaignController.php#L79)

In a lot of our controllers we use the same schema both for validating the parameters and also for sanitizing the response. Either one could work here, but some of the schema needs to be adjusted to reflect how it's validated and how it's sanitized.

ianlin replied:
Thanks for the info. Also updated in eb0425b to use a new schema for query parameters, and to use `get_params` instead of `get_query_params` to get the `country_codes` parameter.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
+					],
+					400
+				);
+			}
```
I think we should leave this to just validate the currency since that's an internal value. We should validate the country code using the built in Rest API validation.

ianlin replied:
Thanks for the info. Also updated in eb0425b to use a new schema for validating and sanitising query parameters, and remove the `country_codes` validation in the controller.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
+			$recommendations = $this
 				->budget_recommendation_query
-				->where( 'country', $country )
+				->where( 'country', explode( ',', $country_codes ), 'IN' )
```
I think it would be better to pass country codes as an array instead of a comma separated list.  Search for the function `get_collection_params` to see some examples of how we pass parameters in the query. If we setup the validation the same then it will be handled for us and we don't need to do an explode here.

ianlin replied:
Thanks for the info. Also updated in eb0425b to get the `country_codes` array instead of a comma separated string.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -68,31 +68,60 @@ public function register_routes(): void {
+							return (float) $recommendation['daily_budget_high'];
+						},
+						$recommendations
+					),
```
I think this structure isn't as clear as it should be because it results in the following:
```json
{
	"currency": "EUR",
	"country_codes": [
		"US",
		"GB",
		"IE"
	],
	"daily_budget_low": [
		10,
		5,
		5
	],
	"daily_budget_high": [
		25,
		20,
		15
	]
}
```

Which means we have to assume that all arrays are in the same order and we need to use the country_codes array as the index for the other two. As far as I recall we've stuck to a pattern like this for most of our controllers which return multiple results:
```json
{
	"currency": "EUR",
	"recommendations": [
		{
			"country": "US",
			"daily_budget_low": 10,
			"daily_budget_high": 25
		},
		{
			"country": "GB",
			"daily_budget_low": 5,
			"daily_budget_high": 20
		},
		{
			"country": "IE",
			"daily_budget_low": 5,
			"daily_budget_high": 15
		}
	]
}
```

ianlin replied:
Good point, the new format is much clearer. Fixed in 07669b2.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -113,24 +142,24 @@ protected function get_schema_properties(): array {
+				'description'       => __( 'An array of the lower limit for the recommended budget.', 'google-listings-and-ads' ),
 				'context'           => [ 'view' ],
 				'validate_callback' => 'rest_validate_request_arg',
 				'required'          => true,
```
I realize these were already here, but if we aren't validating then these won't do anything. And if we do turn on validation, this will fail because `required = true` means it needs to always be passed as a parameter. If we are going to use the same schema for parameters we could set `readonly = true` but I think it would be better to have different schema for the parameters.

ianlin replied:
Thanks for the info. Also updated in eb0425b to use a new schema for validating the parameters.

mikkamp commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -0,0 +1,180 @@
+
+		$this->budget_recommendation_query->expects( $this->exactly(2) )
+			->method( 'where' )
+			->willReturn( $this->budget_recommendation_query );
```
I haven't seen any good examples of mocking chained functions, but I think it would look a little cleaner if we just move this to the setUp function and have it always return itself:

```php
$this->budget_recommendation_query->method( 'where' )->willReturn( $this->budget_recommendation_query );
```

ianlin replied:
Indeed, this is much cleaner. Updated in 4aff2bb to move the mocking `where` clause to the setUp function.

mikkamp commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -0,0 +1,180 @@
+		parent::setUp();
+
+		$this->budget_recommendation_query = $this->createMock( BudgetRecommendationQuery::class );
+		$this->iso_provider = $this->createMock( ISO3166DataProvider::class );
```
If we get the validation implemented then we can change the test `test_get_budget_recommendation_with_nonexistent_country_code` to have this provider return false instead of an empty result. Otherwise we aren't using this mocked class.

ianlin replied:
Thanks for spotting this! Updated in eb0425b to have the iso provider throws an exception for testing.

mikkamp reviewed (commented):
Thanks for the PR, it looks good and it returns the expected results.

However I left some comments on things I think we can refactor to stick a bit closer to the WP REST API standards.

ianlin committed:
Change response payload to be clearer for budget recommendation API

ianlin commented:
Thanks @mikkamp, they are very valuable comments! I'm adjusting the code based on your suggestions above and I will let  you know when I finish it.

ianlin committed:
Update country_codes param form from comma separated string to array

So we don't need to explode the comma-separated string in the
controller.

- Add schema for parameters for budget recommendation API.
- Use `get_collection_params` and validate/sanitize for the
	country_codes parameter.
- Only valide currency in the controller since country_codes are
	verified by the built-in REST validation mechanism.

ianlin committed:
Move mocking `where` method to the setUp function

ianlin committed:
Fix unit test error when testing invalid country

ianlin commented:
Hey @mikkamp, I've done updating the code based on your suggestions, please review it again when you're available, thanks!

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
-			$country        = strtoupper( $request->get_param( 'country_code' ) );
-			$currency       = $this->middleware->get_ads_currency();
-			$recommendation = $this
+			$country_codes = $request->get_params()['country_codes'];
```
Why not just `get_param`? One of the warnings that appears is `Undefined index: country_codes` if we send a request without country_codes (see comment on required collection parameter).

```suggestion
			$country_codes = $request->get_param( 'country_codes' );
```

ianlin replied:
Ah my bad, I replaced it from `get_query_params` and kept the `params` so it became `get_params` 😓. I should check the log messages to see if there are any warnings. Thanks for pointing this out!

Updated in 37e5212.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
+				'items'             => [
+					'type' => 'string',
+				],
+			],
```
Why not set this to `required => true`? I get several PHP warnings when I call the request without `country_codes`, setting it to required will prevent that from happening and show a standard error.

ianlin replied:
Thanks! Also updated in 37e5212.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
+			if ( ! $currency ) {
+				return new Response(
+					[
+						'message'       => __( 'Invalid currency', 'google-listings-and-ads' ),
```
I'm wondering how helpful this error is. Looking at the function `get_ads_currency` it seems it will always return something and if not it will default to the WooCommerce currency. So checking for a "false" value of currency doesn't seem like it would happen. And if it would (because in theory it's still possible to have it return a false value), then we should say that the error is something along the lines of "No currency available for the Ads account", we don't know whether it's a valid/invalid currency.

ianlin replied:
Agree. Updated in eb200bb, thanks!

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -52,47 +52,89 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
+						'currency'      => $currency,
+						'country_codes' => $country_codes,
 					],
 					400
```
If we specifically know it's going to be a "not found" error then it's probably better to use 404 as the status code.

ianlin replied:
Absolutely right! Updated in 57b73b9. I should take more care on it, this was one of my interview questions for the candidates in my previous job. 😅

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -106,34 +148,38 @@ protected function get_budget_recommendation_callback(): callable {
+						],
+						'daily_budget_low'  => [
+							'type'        => 'number',
+							'description' => __( 'Daily budget lower bound ', 'google-listings-and-ads' ),
```
Since it's the field description we can use a full sentence here and end it with a full stop.

ianlin replied:
Thanks and noted! Updated in 7562c39.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -106,34 +148,38 @@ protected function get_budget_recommendation_callback(): callable {
+							'context'           => [ 'view' ],
+							'sanitize_callback' => $this->get_country_code_sanitize_callback(),
+							'validate_callback' => $this->get_country_code_validate_callback(),
+							'required'          => true,
```
The sanitation and validation is important when the schema is used for editing/creating, but here it doesn't really do much, I think just having them in the collection_params function is enough. Also required indicates which fields should be present in a request, so I think we can remove those from all the fields here.

ianlin replied:
Thanks and noted again! Updated in 5a59ebd.

mikkamp commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -0,0 +1,166 @@
+			->method( 'get_ads_currency' )
+			->willReturn( 'TWD' );
+
+		$this->budget_recommendation_query->expects( $this->exactly(2) )->method( 'where' );
```
Is this assertion really useful? We aren't testing the query since it's returning mocked results, so in my opinion it would make sense to do assertions on what we are expecting as the controller response instead.
In my opinion it would only be helpful to check if we are confirming that the the right parameters were sent to the `where` function for each call. But that might be a bit out of scope for this test and better belong in a separate test where the query class is tested.

ianlin replied:
Agree. Updated in a9ce1e2.

mikkamp commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -0,0 +1,166 @@
+		$this->assertEquals( 200, $response->get_status() );
+	}
+
+	public function test_get_budget_recommendation_without_query_parameters() {
```
It looks more like we are testing with an empty string instead of without query parameters. If I test the same in postman:

`GET https://domain.test/wp-json/wc/gla/ads/campaigns/budget-recommendation?country_codes=`
I get the error 'Not a valid alpha2 key'

I guess that's because the iso provider is mocked so it doesn't throw the validation error in the test. But it's probably better to focus on the error we expect without sending any parameters (after setting country_codes to required).

If we want to specifically test for the response `Cannot find any budget recommendations`, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

ianlin replied:
> It looks more like we are testing with an empty string instead of without query parameters.

Thanks for pointing this out! I updated it in 37e5212, I think this is because I did not set `required => true` in the collection params schema so the built-in validator didn't respond the error 400, also as you said the iso provider was mocked so it didn't respond the correct "invalid alpha2 key" error in the unit test.

> If we want to specifically test for the response Cannot find any budget recommendations, I think we should create a separate test where we do pass the countries in the request but get an empty response from the query. A scenario like that could happen if the recommendations DB import failed.

For this I updated in 57b73b9.

mikkamp reviewed (commented):
Thanks for all the changes, those responses look a lot cleaner that way.

I left some additional comments, it doesn't seem the unit tests are totally lined up with what we actually receive in invalid requests, so it's probably better to make that a bit more similar so we can catch any potential errors with the unit tests.

ianlin committed:
Handle no parameters provided error properly for budget recommendation
API

- Use `get_param` instead of `get_params` and add `requred => true` to
	`country_codes` in collection params schema in otder to remove the PHP
  warnings when no parameters are provided.
- Rewrite the unit test to test when no parameters are provided
	correctly.

ianlin committed:
Add a test case when it cannot find any budget recommendations

- Also update the HTTP response status to 404 as it is not found.

ianlin committed:
Update the error message when the currency is a false value

ianlin committed:
Remove sanitize/validate callback and the required in response schema

ianlin committed:
Use full sentence with a full stop in field description of daily budget

ianlin committed:
Remove an unnecessary assertion in budget recommendation tests

ianlin commented:
Hello @mikkamp, thanks for all the comments you gave, they are really useful. I've updated the code based on your suggestions, please have it a look again when you're available.

mikkamp commented on src/API/Site/Controllers/Ads/BudgetRecommendationController.php:
```diff
@@ -52,47 +52,90 @@ public function __construct( RESTServer $rest_server, BudgetRecommendationQuery
+				function ( $recommendation ) {
+					return [
+						'country'           => $recommendation['country'],
+						'daily_budget_low'  => (float) $recommendation['daily_budget_low'],
+						'daily_budget_high' => (float) $recommendation['daily_budget_high'],
```
Do we need these to be cast to floats? In the database the columns are defined as `int(11)` so seems like casting to an int would do here.

mikkamp commented on tests/Unit/API/Site/Controllers/Ads/BudgetRecommendationControllerTest.php:
```diff
@@ -0,0 +1,178 @@
+			'recommendations' => [
+				[
+					'country'           => 'US',
+					'daily_budget_low'  => 330.0,
+					'daily_budget_high' => 930.0,
```
Even though we cast to float I seem to always be getting these values as 330 or 930, but never with the decimal added. Can we drop the decimals since it doesn't get returned that way in a response? It seems they are only added in the test since we are comparing with assertSame (which is strict about comparing the type), but if we cast them to int that's no longer needed.
This is what I'm seeing in the response even if they are floats, so I'd expect us to be testing with similar values:
```json
{
	"country": "US",
	"daily_budget_low": 10,
	"daily_budget_high": 25
}
```

mikkamp reviewed (approved):
Thanks for all those extra changes, it's working well, and I'm getting the expected results. So I'll go ahead and mark it as approved.

Tested:
- unit tests
- response with invalid parameters
- response with multiple country code
- response with invalid country code

I just left two small comments on the numbers being float values, doesn't seem like that's necessary and it's a bit confusing in the test since we aren't actually getting them with the decimals in the response.

ianlin committed:
Cast the budget value to int rather than float
